### PR TITLE
[aot autograd] refactor CompiledFunction.backward: control flow (3/N)

### PR DESCRIPTION
**Pull request by xmfan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #139377
* __->__ #139347
* #139343
* #139331

### CODE REVIEW

**Review by zou3519 — APPROVED** [review]

File: torch/_functorch/_aot_autograd/runtime_wrappers.py [hunk 1]
```diff
@@ -1647,7 +1647,46 @@ def forward(ctx, *deduped_flat_tensor_args):
 
             @staticmethod
             def backward(ctx, *flat_args):
-                return CompiledFunction._backward_prologue(ctx, *flat_args)
+                all_args = CompiledFunction._backward_prologue(ctx, *flat_args)
+
+                def impl_fn(double_ctx=None):
+                    out = CompiledFunction._backward_impl(ctx, all_args)
```
Comment (bdhirsh) [inline torch/_functorch/_aot_autograd/runtime_wrappers.py:1653]: is the plan eventually for CA to inline these 3 calls into the CA graph, and for dynamo to further inline `_backward_impl` (to handle any backward hooks that are inside of the `bw_module`?

Comment (xmfan) [reply torch/_functorch/_aot_autograd/runtime_wrappers.py:1653]: the prologue and epilogue aren't very friendly to inline, we're just planning to allow_in_graph them

other than that, yes, the aot backward graph called by the `_backward_impl` CA codepath needs to be inlined to inspect the bwd hooks

Comment (xmfan) [reply torch/_functorch/_aot_autograd/runtime_wrappers.py:1653]: yeah we need to make some changes in dynamo in the next PR for 
(1) and change it to `in_compiled_autograd_region`
(2) i think we can route it via FakeBackwardCFunction (for the make_fx CA at least, need to double check for python CA)

**Pull request closed without merging**